### ISSUE TITLE
NaN in  frame.IIII.position.hdf5

### ISSUE THREAD

**Issue by Matstah** [apple/ml-hypersim#13]
Hi there

I am using `ai_035_001/images/scene_cam_00_geometry_hdf5/frame.0001.position.hdf5` to get the world positions of all pixels seen on the image.

When looking at pixel (257,837), and printing the 3D point by `print('xyz_world: ', xyz_world[257,837,:])`, I get a NaN vector back `xyz_world:  [nan nan nan]`.

Is there any meaning behind the nan value used for a xyz world point?

Thanks

**Comment by mikeroberts3000** [comment]
Hi! Great question. NaN values can and do occur in the Hypersim images when there is no scene geometry at that pixel. In the case of `{ai_035_001, cam_00, frame.0001}`, we see that there are indeed some "empty" pixels, i.e., pixels that don't observe any scene geometry. Here are some preview images from this frame.

`frame.0001.tonemap.jpg`

![frame 0001 tonemap](https://user-images.githubusercontent.com/2341965/105396959-565c8e00-5bd5-11eb-96c2-738193eefc72.jpg)

`frame.0001.depth_meters.png`

![frame 0001 depth_meters](https://user-images.githubusercontent.com/2341965/105397015-64aaaa00-5bd5-11eb-8389-88a5f07913d3.png)

White values in the `frame.0001.depth_meters.png` preview image indicates empty pixels, and we expect to have NaN values in these pixels in our HDF5 files. The exact pixel coordinate you're referring to appears to correspond to the top-right pixel of the right-most window.

In summary, you're seeing a NaN value because there is no geometry at that pixel, and therefore the pixel's world-space position is undefined.

**Comment by rikba** [comment]
We also came across this issue when we tried to use the [Evermotion data set in a multi view geometry application](https://github.com/rikba/hypersim_multiview). 
For example in volume 44, scene 2 some pixel blobs in the depth image are not defined:
![hole_in_depths](https://user-images.githubusercontent.com/11293852/121885085-06ad6900-cd14-11eb-9f81-423acfa97fca.png)

Quite many scenes have small portions that have undefined geometry information. When trying to backproject all pixels `p` of one frame into its own camera plane using the pixel world position `r_WP`,  camera intrinsics `K`, and camera pose `T_CW`, i.e., 
```
p_projected = K * T_CW * r_WP(p_original)
```
we'd expect  `p_projected == p_original`. The histogram below shows the maximum reprojection error for every frame. We suspect that any image that has a maximum reprojection error greater than `100` pixel contains missing geometry information.

![max_error](https://user-images.githubusercontent.com/11293852/121885871-f8ac1800-cd14-11eb-8dd7-92bcbf62080b.png)

We did not cross-compare whether huge reprojection errors are always caused by undefined geometry information.
However, we made the [frame-to-itself reprojection results available](https://github.com/rikba/hypersim_multiview/blob/bf1d42bedcbd541ee98435626db8de7c68d6ad16/results/error_statistics.csv).
Thus users can possibly remove frames or mask frame patches where the geometry information is problematic.

**Comment by mikeroberts3000** [comment]
@rikba this analysis is great!

I could imagine that 30k images (roughly half of the images in Hypersim) contain at least a few pixels with undefined geometry. As we see in the original post above, all it takes is for there to be a visible open window in an otherwise well-behaved image, and you'll get some undefined values. So 30k seems plausible to me. As you already mentioned, it would be interesting to see how many of those >100px max errors are due to undefined geometry. Is that information stored in your error statistics CSV file? Or are the NaN values being converted to finite floating-point values somehow by the time they are saved to your CSV. I hope it is the former.

The other (finite) reprojection errors tell us exactly what scenes are affected by https://github.com/apple/ml-hypersim/issues/24. This analysis is very valuable! Thank you 😄

**Comment by rikba** [comment]
> Is that information stored in your error statistics CSV file? Or are the NaN values being converted to finite floating-point values somehow by the time they are saved to your CSV. I hope it is the former.

Unfortunately, the latter. We did not check the values in `frame.FFFF.position.hdf5` before doing the reprojection. Thus I think the computation assumes some random memory for the world position of the pixel and comes up with a random reprojection error, e.g. `1.3043818e+19` pixel that is stored in the CSV. But since this evaluation seems valueable, we might rerun it and just flag pixel that are undefined.
